### PR TITLE
Copy Gromacs from bioconda

### PR DESCRIPTION
**Pull request by jan-janssen**
Original: https://github.com/bioconda/bioconda-recipes/tree/master/recipes/gromacs/2021 

Copied primarily to integrate with https://github.com/conda-forge/staged-recipes/pull/17176 and https://github.com/conda-forge/plumed-feedstock 

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by bgruening** [comment]
ping @simonbray

**Comment by simonbray** [comment]
ping @mabraham

**Comment by bgruening** [comment]
See also: https://github.com/bioconda/bioconda-recipes/pull/31508

File: recipes/gromacs/meta.yaml
Comment (bgruening) [inline recipes/gromacs/meta.yaml:75]: You should probably add the other maintainers from bioconda as well (gi history) or ask them.

File: recipes/gromacs/build.sh
Comment (mabraham) [inline recipes/gromacs/build.sh:28]: Should change to conda

File: recipes/gromacs/build.sh
Comment (mabraham) [inline recipes/gromacs/build.sh:13]: In Bioconda, CI time was limiting, so we opted not to build an AVX_512 flavor. We could consider relaxing that here. But before we do so, we should also consider whether we plan to add a CUDA build. (I forget why that wasn't done in bioconda, but perhaps the reason was that the CUDA support was not as good as in conda-forge.)

**Comment by jan-janssen** [comment]
I was not aware of https://github.com/bioconda/bioconda-recipes/pull/31676 but that is basically what I intended as well. My general issue was that I tried to create a conda package for https://www.votca.org - https://github.com/conda-forge/staged-recipes/pull/17176 and I was not able to link from conda-forge to bioconda while linking from bioconda to conda-forge seems to be fine. I asked about this on the conda-forge Gitter channel and the suggestion was to open a pull request with the copied recipe to discuss it. If there is a way to link bioconda dependencies to conda-forge packages that is fine for me, otherwise I would prefer to migrate gromacs to conda-forge. My personal background is more in atomistic material science simulation, so I created conda-forge packages for MD codes like https://github.com/conda-forge/lammps-feedstock and DFT codes like https://github.com/conda-forge/qe-feedstock https://github.com/conda-forge/siesta-feedstock https://github.com/conda-forge/abinit-feedstock https://github.com/conda-forge/cp2k-feedstock . I use these packages in my own software to build simulation workflows https://pyiron.org . Finally for my current post-doc I work more in chemistry hence the interest in votca. I find it hard to draw a line between bioconda and conda-forge but if there is a strong preference not to migrate gromacs I can understand that as well.

File: recipes/gromacs/build.sh
Comment (douglowe) [reply recipes/gromacs/build.sh:13]: I believe it was the lack (at the time) of CUDA support in the bioconda build system which did stop us adding that support.

**Comment by simonbray** [comment]
As a user of the bioconda GROMACS package, who also has contributed packages to bioconda and CF, I would also be in favour of the migration for the following reasons (some of which @jan-janssen already mentioned):

- ideally builds for CUDA and OpenCL can be added to conda at some point, and my understanding is that the support for that is better on CF than bioconda
- bioconda is intended as a specialized channel for software with a biological application, especially bioinformatics, and GROMACS has a broader user base
- as a consequence of the previous point, bioconda supports CF dependencies, but the reverse is not true. So the result is that any conda packages which use GROMACS as a dependency cannot currently be added to CF, but only to bioconda. Moving to CF would give more flexibility to developers of downstream tools.

**Comment by simonbray** [comment]
> a potential channel prioritization issues if we don't deprecate the packaged named the same bioconda side?

I think the bioconda one would be removed - at least, that was my experience when migrating AutoDock Vina: https://github.com/bioconda/bioconda-recipes/pull/30856

**Comment by douglowe** [comment]
[bioconda/bioconda-recipes#31676](https://github.com/bioconda/bioconda-recipes/pull/31676) is at a very early stage of development - I'm quite happy to help with the development of a conda-forge package instead, if that is the route we take.

To continue with using AutoDock Vina as an example of a transition, it seems there was a package name change too, but older versions of the autodock-vina packages remain available via bioconda:
```
# Name                       Version           Build  Channel             
autodock-vina                  1.1.2               0  bioconda            
autodock-vina                  1.1.2               1  bioconda            
autodock-vina                  1.1.2               2  bioconda            
# Name                       Version           Build  Channel             
vina                           1.2.2 py310he60a7ad_1  conda-forge         
vina                           1.2.2  py37h48520ad_0  conda-forge         
vina                           1.2.2  py37h48520ad_1  conda-forge         
vina                           1.2.2  py38h46efa4d_0  conda-forge         
vina                           1.2.2  py38h46efa4d_1  conda-forge         
vina                           1.2.2  py39h03712de_0  conda-forge         
vina                           1.2.2  py39h03712de_1  conda-forge         
vina                           1.2.3 py310he60a7ad_0  conda-forge         
vina                           1.2.3  py37h48520ad_0  conda-forge         
vina                           1.2.3  py38h46efa4d_0  conda-forge         
vina                           1.2.3  py39h03712de_0  conda-forge 
```

Perhaps a shift to conda-forge could be carried out for the next big gromacs version change (2022) - with 2021, and earlier, versions remaining on bioconda?

File: recipes/gromacs/meta.yaml
Comment (douglowe) [inline recipes/gromacs/meta.yaml:4]: Can you try version 2021.4 instead?

**Comment by simonbray** [comment]
> To continue with using AutoDock Vina as an example of a transition, it seems there was a package name change too, but older versions of the autodock-vina packages remain available via bioconda

Yes, to clarify: the recipe was removed from the bioconda GitHub repo, but the older builds were not deleted. (The rename was just for consistency with the `vina` PyPI package, which didn't exist when the bioconda recipe was first written).

**Comment by jan-janssen** [comment]
@simonbray @douglowe  @mabraham @stain If you do not mind I am going to add you as maintainers to the recipe. Can you give me a quick *thumbs up* if you agree?

File: recipes/gromacs/build.sh
Comment (jan-janssen) [reply recipes/gromacs/build.sh]: Done

File: recipes/gromacs/meta.yaml
Comment (jan-janssen) [reply recipes/gromacs/meta.yaml]: Done

**Comment by jan-janssen** [comment]
@conda-forge/staged-recipes, ready for review

File: recipes/gromacs/build.sh
Comment (junghans) [inline recipes/gromacs/build.sh:35]:
```suggestion
      cmake_args+=(-DGMX_MPI=OFF -DGMX_THREAD_MPI=ON)
```
So one at least gets thread-mpi version of gromacs.

File: recipes/gromacs/meta.yaml
Comment (jan-janssen) [reply recipes/gromacs/meta.yaml:75]: Done

**Comment by bgruening** [comment]
What Simon said :-)
Bioconda depends on CF, but not the other way around. Also bioconda recommends to have CF with a higher priority than bioconda. This means we will always get the latest builds from CF until we specify a concrete older version ( from the bioconda channel).

Thanks all!

**Comment by douglowe** [comment]
I think it makes sense to have a clean break between the conda-forge and bioconda builds - so start the conda-forge provision from version 2022 onwards (and keep versions up to 2021 on bioconda). There's also a few improvements that I think we should make to the package before publication. I've started an issues thread (https://github.com/conda-forge/staged-recipes/issues/17203) for discussing how we want to handle these.

File: recipes/gromacs/build.sh
Comment (mabraham) [reply recipes/gromacs/build.sh:13]: Ok. Nothing *needs* to change here, but AVX_512 is desirable at some future time

File: recipes/gromacs/build.sh
Comment (mabraham) [inline recipes/gromacs/build.sh:74]: Perhaps using the `cat`-based approach below is one that avoids having to learn how to access the directory from which this script was read, because that need not relate to the current working directory at all.

Comment (douglowe) [reply recipes/gromacs/build.sh:74]: Yes, I was just thinking that was (another) benefit of using the `cat` approach.

The other option would be to put the `cp` command in the yaml build file - as we did for the prismatic feedstock: https://github.com/conda-forge/prismatic_split-feedstock/blob/a41dd1ef0330cdfb7e09aced025a64822ed1df50/recipe/meta.yaml#L152-L155

Comment (douglowe) [reply recipes/gromacs/build.sh:74]: Well, not quite the same, but we could probably have some commands to copy the deactivate scripts across.

Comment (douglowe) [reply recipes/gromacs/build.sh:74]: Ahh, no - we could use this instead:
```
cp "${RECIPE_DIR}/gromacs_deactivate.sh" "${PREFIX}/etc/conda/deactivate.d/"
```
That should work.

Comment (jan-janssen) [reply recipes/gromacs/build.sh]: Thanks - this worked.

**Comment by mabraham** [comment]
I think the work is done and we can return to this in January when the 2022 release is made. Is there anything else we should do to prepare to use conda forge for future GROMACS versions?

**Comment by douglowe** [comment]
The OSX and Win_64 builds are failing with the error:
`conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {MatchSpec("django-keyedcache3")}`

I don't think this is a GROMACS dependency? I wonder if it's a conda-forge build dependency.

@jan-janssen - could you make a trivial change, and force the build process to try again? See if this problem goes away...

**Comment by mabraham** [comment]
> I don't think this is a GROMACS dependency?

It certainly isn't, unless somehow it's a dependency of a dependency...

File: recipes/gromacs/meta.yaml
Comment (douglowe) [inline recipes/gromacs/meta.yaml:5]: Even though the gromacs tar file is labelled 2020, should we still set the version here to 2020.0 (and hard code the url for this download)?

Comment (douglowe) [reply recipes/gromacs/meta.yaml:5]: No - ignore me - we have followed this convention in the bioconda releases too.

**Comment by douglowe** [comment]
@conda-forge/staged-recipes - please could you review this PR? All checks have been successfully passed.

**Comment by jan-janssen** [comment]
@bgruening Can you maybe review this pull request?

**Comment by dpryan79** [comment]
This is fine from the bioconda side, we can just remove the recipe from our repo after merging.

**Review by bgruening — APPROVED** [review]
this looks great to me! Thanks everyone!

**Comment by simonbray** [comment]
Just installed 2022, so nice to have it in conda-forge - thanks @jan-janssen and everyone else for working on it!

**Pull request merged**
